framework/src/modules/poa/module.ts

```diff
@@ -31,11 +42,183 @@ export class PoAModule extends BaseModule {
+				throw new Error('`name` property is invalid. Must contain only characters a-z0-9!@$&_.');
+			}
+		}
+
+		const { activeValidators, threshold } = snapshotSubstore;
+		const activeValidatorAddresses = activeValidators.map(v => v.address);
+		const sortedActiveValidatorsByAddress = [...activeValidatorAddresses].sort((a, b) =>
+			a.compare(b),
+		);
+		const validatorAddressesString = validatorAddresses.map(a => a.toString('hex'));
+		let totalWeight = BigInt(0);
+
+		// Check that the address properties of entries in the snapshotSubstore.activeValidators are pairwise distinct.
+		if (!objects.bufferArrayUniqueItems(activeValidatorAddresses)) {
+			throw new Error('`address` properties in `activeValidators` must be distinct.');
+		}
+
+		for (let i = 0; i < activeValidators.length; i += 1) {
+			// Check that entries in the snapshotSubstore.activeValidators array are ordered lexicographically according to address.
+			if (!activeValidators[i].address.equals(sortedActiveValidatorsByAddress[i])) {
+				throw new Error(
+					'`activeValidators` must be ordered lexicographically by address property.',
+				);
+			}
+
+			// Check that for every element activeValidator in the snapshotSubstore.activeValidators array, there is an entry validator in the validators array with validator.address == activeValidator.address.
+			if (!validatorAddressesString.includes(activeValidators[i].address.toString('hex'))) {
+				throw new Error('`activeValidator` address is missing from validators array.');
+			}
+
+			// Check that the weight property of every entry in the snapshotSubstore.activeValidators array is a positive integer.
+			if (activeValidators[i].weight <= BigInt(0)) {
+				throw new Error('`activeValidators` weight must be positive integer.');
+			}
+
+			totalWeight += activeValidators[i].weight;
+		}
+
+		if (totalWeight > MAX_UINT64) {
+			throw new Error('Total weight `activeValidators` exceeds maximum value.');
+		}
+
+		// Check that the value of snapshotSubstore.threshold is within range
+		if (threshold < totalWeight / BigInt(3) + BigInt(1) || threshold > totalWeight) {
+			throw new Error('`threshold` in snapshot substore is not within range.');
+		}
+
+		// Create an entry in the validator substore for each entry validator in the validators
+		// Create an entry in the name substore for each entry validator in the validators
+		const validatorStore = this.stores.get(ValidatorStore);
+		const nameStore = this.stores.get(NameStore);
+
+		for (const currentValidator of validators) {
+			await validatorStore.set(context, currentValidator.address, { name: currentValidator.name });
+			await nameStore.set(context, Buffer.from(currentValidator.name, 'utf-8'), {
+				address: currentValidator.address,
+			});
+		}
+
+		// Create three entries in the snapshot substore indicating a snapshot of the next rounds of validators
+		const snapshotStore = this.stores.get(SnapshotStore);
+		await snapshotStore.set(context, utils.intToBuffer(0, 4), {
```

Comment:
You can use constant here for snapshot keys which were introduced in https://github.com/LiskHQ/lisk-sdk/pull/8523. Also in other places

https://github.com/LiskHQ/lisk-sdk/blob/d109544c6e1c937a07a1778b6f2e226675dbde8d/framework/src/modules/poa/constants.ts#L40-L44